src/js/selection/SelectionManager.js

```diff
@@ -101,14 +107,36 @@
 
     this.pastePixels_(frame, pixels);
 
-    $.publish(Events.PISKEL_SAVE_STATE, {
-      type : pskl.service.HistoryService.REPLAY,
-      scope : this,
-      replay : {
-        type : SELECTION_REPLAY.PASTE,
-        pixels : JSON.parse(JSON.stringify(pixels.slice(0)))
+    // Offset the pasted selection from the original location.
+    var tool = pskl.app.drawingController.currentToolBehavior;
+    var isSelectionTool = tool instanceof pskl.tools.drawing.selection.BaseSelect;
+    if (isSelectionTool) {
+      var maxHeight = frame.width - 1;
+      var maxWidth = frame.height - 1;
+      var xCoordinate = pixels[pixels.length - 1].col;
+      var yCoordinate = pixels[pixels.length - 1].row;
+      if (xCoordinate < maxWidth) {
+        this.currentSelection.move(offset, 0);
       }
-    });
+      if (yCoordinate < maxHeight) {
+        this.currentSelection.move(0, offset);
+      }
+
+      var overlay = pskl.app.drawingController.overlayFrame;
+      tool.reDraw(overlay);
+    }
+
+    // Save to state when selection is dropped in place.
+    if (offset === 0) {
```

Comment:
Discussed in person: Maybe should create a save state for each paste operation _after_ the first one and the deselect operation, instead of only on the deselect operation?